### PR TITLE
Added git contribution report

### PR DESCRIPTION
**Pull request by shivanshuag**
Issue #56 
- Gives a list of all the git contributors along with lines added, lines deleted and percentage contribution.
- Percentage contribution is calculated by adding lines added and lines deleted and dividing them by total number of lines added + total number of lines deleted.

```
 ❯ drush acb --detail                                                     [17:38:23]
Codebase: Info
  Size of public:// folder: Determine the size of public:// folder.
    Files: 3.26MB
  Size of entire site: Determine the size of the site root; does not include remote mounts.
    Total size: 168.77MB
  Drupal managed file count: Determine the count of Drupal managed files.
    Managed file count: 0
  Drupal managed file size: Determine the size of Drupal managed files.
    Managed file size: 0.00kB
  Git Contributors: Determine lines of code added and removed by each contributor.
    Author: Inserted, Deleted, Percentage
    -----------------------------
    webchick: 1562040, 1067003, 21.08
    Alex Pott: 1107628, 788839, 15.21
    catch: 1294634, 519150, 14.54
    Dries: 970634, 498109, 11.78
    Nathaniel: 928120, 283966, 9.72
    Dries Buytaert: 604622, 376895, 7.87
    Nathaniel Catchpole: 391203, 277588, 5.36
    Nathan Haug: 264364, 264341, 4.24
    Angie Byron: 170674, 85308, 2.05
    git: 88566, 76952, 1.33
    Tim Plunkett: 72229, 67726, 1.12
    Larry Garfield: 43103, 36023, 0.63
    merlinofchaos: 77303, 0, 0.62
    Daniel Wehner: 31977, 31661, 0.51
    damiankloip: 32296, 27538, 0.48
    Steven Wittens: 38753, 20205, 0.47
    effulgentsia: 18330, 37774, 0.45
    Gábor Hojtsy: 31268, 18803, 0.40
    Jennifer Hodgdon: 27383, 18615, 0.37
    Kjartan Mannes: 17739, 14897, 0.26
    dereine: 11592, 6790, 0.15
    xjm: 9410, 5757, 0.12
    James Gilliland: 13899, 113, 0.11
    Neil Drumm: 7231, 4032, 0.09
    Jeroen Bensch: 6212, 3680, 0.08
    Andrew Berry: 6794, 1924, 0.07
    Gerhard Killesreiter: 4764, 3550, 0.07
    Dave Reid: 4332, 4143, 0.07
    Damian Lee: 9263, 61, 0.07
    amateescu: 4026, 3871, 0.06
    berdir: 3819, 3819, 0.06
    tim.plunkett: 3282, 3286, 0.05
    gdd: 3190, 2113, 0.04
    Greg Dunlap: 3179, 2110, 0.04
    robloach: 2636, 2569, 0.04
    pl2: 2622, 2576, 0.04
    Andreas Hennings: 2651, 2201, 0.04
    chrisjlee: 2308, 2308, 0.04
    pablo.cerda: 1613, 1570, 0.03
    Bram Goffings: 1681, 1849, 0.03
    Katherine Bailey: 1754, 961, 0.02
    bojanz: 1222, 980, 0.02
    ayelet_Cr: 2435, 374, 0.02
    natrak: 1100, 1207, 0.02
    aspilicious: 1461, 1449, 0.02
    karst: 510, 497, 0.01
    Niklas Fiekas: 384, 251, 0.01
    slv: 345, 433, 0.01
    sun: 571, 463, 0.01
    alasdair86: 826, 670, 0.01
    Bojan Zivanovic: 723, 898, 0.01
    The Great Git Migration: 815, 0, 0.01
    osopolar: 2, 3, 0.00
    palagyi: 67, 100, 0.00
    patdunlavey: 1, 1, 0.00
    nod_: 19, 9, 0.00
    msonnabaum: 90, 20, 0.00
    anonymous: 63, 0, 0.00
    andypost: 3, 3, 0.00
    slashrsm: 30, 12, 0.00
    aaronbauman: 4, 1, 0.00
    StoraH: 1, 1, 0.00
    yan_nick: 20, 7, 0.00
    yasha: 16, 0, 0.00
    zuuperman: 9, 4, 0.00
    yorirou: 2, 3, 0.00
    sphism: 1, 1, 0.00
    matslats: 6, 3, 0.00
    Pierre Rineau: 8, 6, 0.00
    Peter Drake: 86, 40, 0.00
    Pol: 21, 0, 0.00
    recrit: 2, 2, 0.00
    rvilar: 10, 0, 0.00
    rocket: 2, 2, 0.00
    Pedro Lozano: 182, 179, 0.00
    Kris: 37, 26, 0.00
    Gábor Hojtsy: 11, 3, 0.00
    fago: 191, 74, 0.00
    enzo - Eduardo Garcia: 2, 2, 0.00
    Gabor Hojtsy: 1, 1, 0.00
    GK: 10, 0, 0.00
    Ignigena: 1, 1, 0.00
    hass: 7, 2, 0.00
    greggles: 4, 0, 0.00
    drothstein: 35, 11, 0.00
    Chi: 1, 1, 0.00
    daspeter: 12, 2, 0.00
    cosmicdreams: 1, 1, 0.00
    Damien Tournoud: 28, 45, 0.00
    David Strauss: 228, 82, 0.00
    dead_arm: 117, 89, 0.00
    dmitriytrt: 3, 4, 0.00
    Claes Gyllensvärd: 6, 4, 0.00
    jacintocapote: 55, 0, 0.00
    Bob Vincent: 25, 7, 0.00
    dagmar: 3, 1, 0.00
    Katherine Senzee: 86, 10, 0.00
    Karoly Negyesi: 273, 140, 0.00
    larowlan: 162, 108, 0.00
    bellesmanieres: 20, 8, 0.00
    lunaris: 7, 7, 0.00
    Lin Clark: 3, 5, 0.00
    Letharion: 13, 1, 0.00
    Justin Randell: 126, 29, 0.00
    justinrandell: 12, 9, 0.00
    =: 53, 1, 0.00
    blauerberg: 1, 1, 0.00
    jax: 1, 1, 0.00
    jhodgdon: 104, 87, 0.00
    joachim: 16, 1, 0.00
    johnv: 9, 2, 0.00
    Jody Hamilton: 98, 30, 0.00
    Mark Sonnabaum: 29, 0, 0.00
```

### CODE REVIEW

**Comment by fluxsauce** [comment]
Well, that's not useful when it's the Drupal codebase :-/ I have to think about this one.

File: Check/Codebase/GitContributor.php [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+      }
+      $ret_val .= '-----------------------------';
+      foreach ($this->registry['git_contribution_percentage'] as $user => $percentage) {
+        $ret_val .= PHP_EOL;
+        if (!drush_get_option('json')) {
+          $ret_val .= str_repeat(' ', 4);
+        }
+        $added = $this->registry['git_contribution'][$user]['inserted'];
+        $deleted = $this->registry['git_contribution'][$user]['deleted'];
+        $ret_val .= "$user: $added, $deleted, $percentage";
```
Comment (fluxsauce) [inline Check/Codebase/GitContributor.php]: `, $percentage%`

File: site_audit.drush.inc [hunk 2]
```diff
@@ -109,6 +109,7 @@ function site_audit_drush_command() {
       'SizeAll',
       'ManagedFileCount',
       'ManagedFileSize',
+      'GitContributor',
```
Comment (fluxsauce) [inline site_audit.drush.inc]: `GitContributions`

**Comment by fluxsauce** [comment]
Well, this is exactly what I asked for but it's not completely useful in this particular case.

Uh, can you check to see if the git repository is from drupal.org or github.com/drupal/drupal or github.com/pressflow and https://github.com/pantheon-systems/drops-7/ and not run it in those instances?

File: Check/Codebase/GitContributions.php [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+   * Implements \SiteAudit\Check\Abstract\getResultFail().
+   */
+  public function getResultFail() {}
+
+  /**
+   * Implements \SiteAudit\Check\Abstract\getResultInfo().
+   */
+  public function getResultInfo() {
+    if (isset($this->registry['not_git'])) {
+      return dt('Site is not under git source control');
```
Comment (fluxsauce) [inline Check/Codebase/GitContributions.php]: The site is not stored in Git.

File: Check/Codebase/GitContributions.php [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    }
+    if (isset($this->registry['git_remote'])) {
+      return dt("The repository has whole of drupal's git history. Reporting contrinutions of users will not be helpful in this case.");
```
Comment (fluxsauce) [inline Check/Codebase/GitContributions.php]: `return dt('The repository contains the entire Drupal contribution history and will be ignored.');`

**Comment by fluxsauce** [comment]
Does not accurately track deletions.

``` bash
jon@Decker ~/projects/drupal-8.0.0-beta10 (master) $ drush acb --detail
Codebase: Info
  Size of public:// folder: Determine the size of public:// folder.
    Files: 28.46MB
  Size of entire site: Determine the size of the site root; does not include remote mounts.
    Total size: 290.84MB
  Drupal managed file count: Determine the count of Drupal managed files.
    Managed file count: 0
  Drupal managed file size: Determine the size of Drupal managed files.
    Managed file size: 0.00kB
  Git Contributions: Determine lines of code added and removed by each contributor.
    Author: Inserted, Deleted, Percentage
    -----------------------------
    Jon Peck: 2529238, 0, 100.00%
jon@Decker ~/projects/drupal-8.0.0-beta10 (master) $ rm -Rf vendor/drush
jon@Decker ~/projects/drupal-8.0.0-beta10 (master) $ git commit -m "Removing drush"
On branch master
Changes not staged for commit:
    modified:   modules/generate_errors (untracked content)
    deleted:    vendor/drush/drush/.gitignore
    deleted:    vendor/drush/drush/.travis.yml
...
    deleted:    vendor/drush/drush/unish.sh

no changes added to commit
jon@Decker ~/projects/drupal-8.0.0-beta10 (master) $ git add .
jon@Decker ~/projects/drupal-8.0.0-beta10 (master) $ git commit -m "Removing drush"
[master cc5c87c] Removing drush
 319 files changed, 55893 deletions(-)
 delete mode 100644 vendor/drush/drush/.gitignore
 delete mode 100644 vendor/drush/drush/.travis.yml
 delete mode 100644 vendor/drush/drush/CONTRIBUTING.md
 delete mode 100644 vendor/drush/drush/README.md
 delete mode 100644 vendor/drush/drush/commands/core/archive.drush.inc
 delete mode 100644 vendor/drush/drush/commands/core/browse.drush.inc
 delete mode 100644 vendor/drush/drush/commands/core/cache.drush.inc
 delete mode 100644 vendor/drush/drush/commands/core/cli.drush.inc
 delete mode 100644 vendor/drush/drush/commands/core/config.drush.inc
 delete mode 100644 vendor/drush/drush/commands/core/core.drush.inc
 delete mode 100644 vendor/drush/drush/commands/core/docs.drush.inc
...
 delete mode 100755 vendor/drush/drush/unish.sh
jon@Decker ~/projects/drupal-8.0.0-beta10 (master) $ drush acb --detail
Codebase: Info
  Size of public:// folder: Determine the size of public:// folder.
    Files: 28.46MB
  Size of entire site: Determine the size of the site root; does not include remote mounts.
    Total size: 288.06MB
  Drupal managed file count: Determine the count of Drupal managed files.
    Managed file count: 0
  Drupal managed file size: Determine the size of Drupal managed files.
    Managed file size: 0.00kB
  Git Contributions: Determine lines of code added and removed by each contributor.
    Author: Inserted, Deleted, Percentage
    -----------------------------
    Jon Peck: 2585131, 0, 100.00%
jon@Decker ~/projects/drupal-8.0.0-beta10 (master) $ git log
commit cc5c87c9992ea5bf5926e02846a0c4003f869d90
Author: Jon Peck <jpeck@fluxsauce.com>
Date:   Wed Aug 12 13:58:01 2015 -0700

    Removing drush

commit 707296cfd414ef16f084b04169a5db219d3cb99d
Author: Jon Peck <jpeck@fluxsauce.com>
Date:   Wed Aug 12 13:57:09 2015 -0700

    Initial commit.
```

**Comment by fluxsauce** [comment]
Are you testing your code? This straight up doesn't work, please reexamine.

```
jon@Decker ~/projects/drupal-8.0.0-beta10 (master) $ drush acb --detail
sh: -c: line 0: syntax error near unexpected token `('
sh: -c: line 0: `git log --no-merges --shortstat --author 'Jon Peck' 2> /dev/null | grep 'files\? changed' 2> /dev/null | awk {if ($5=="insertions(+)" || $5=="insertion(+)") inserted+=$4; else deleted+=$4; deleted+=$6} END {print inserted, deleted}' 2> /dev/null'
Division by zero GitContributions.php:134                                                                                                          [warning]
Codebase: Info
  Size of public:// folder: Determine the size of public:// folder.
    Files: 28.46MB
  Size of entire site: Determine the size of the site root; does not include remote mounts.
    Total size: 288.09MB
  Drupal managed file count: Determine the count of Drupal managed files.
    Managed file count: 0
  Drupal managed file size: Determine the size of Drupal managed files.
    Managed file size: 0.00kB
  Git Contributions: Determine lines of code added and removed by each contributor.
    Author: Inserted, Deleted, Percentage
    -----------------------------
    Jon Peck: , , 0.00%
```

**Comment by shivanshuag** [comment]
It was really a very silly mistake. There was a single quote missing in the code I committed. I wonder how I missed to commit it, it was there in my local branch. That is why it was working correctly when I tested, and worked correctly on the archive dump you gave me.

**Pull request merged**